include/seastar/http/client.hh

```diff
@@ -230,6 +231,16 @@ public:
      */
     future<> make_request(request req, reply_handler handle, reply::status_type expected = reply::status_type::ok);
 
+    /**
+     * \brief Updates the maximum number of connections a client may have
+     *
+     * If the new limit is less than the amount of connections a client has, they will be
+     * closed. The returned future resolves when all excessive connections get closed
+     *
+     * \param nr -- the new limit on the number of connections
+     */
+    future<> set_maximum_connections(unsigned nr);
```

Comment:
I think it is an overkill for set_maximum_connections() to return a future. It doesn't sound useful to me to wait until the number of connections reduces (maybe I'm missing something?), and it complicates understanding what happens if you, for example, run two of these functions in parallel.
It seems much easier for set_maximum_connections() to just return immediate - just as long as the pool is still longer than the current maximum, we close connections instead of keeping them.
But I don't insist - if you think this is an important feature (and can explain why), you can keep it.

Reply:
Let me know if you want me to commit this version as-is, or if you were convinced by my argument that set_maximum_connections() should be void and I should wait for another version.

Reply:
My first implementation of set_maximum_connections was void and exited immediately just as you suggest. And you're right that void-returning implementation is much simpler. Then I changed my mind, but I don't have strong arguments _for_ the new implementation.

The reason why I changed it was inspired by (SUDDENLY) `file::close()` call which is future-returning (and there was similar discussion around io_intent cancellation, but I cannot find this thread). The rationale behind making it synchronous was to give caller more control over resources. One closes a file and it doesn't detach into background leaving the system with unknown number of not-yet-but-soon-will-be released resources. Same here -- there are potentially N opened sockets owned by the client, top-level decides (whatever it means) that now it can afford only M (M < N) and shrinks the client. Being asynchronous (i.e. -- returning void) the top level doesn't have any control over the amount of resources it actually ... has/maintains/owns, but being synchronous (i.e. -- returning future<>) is more resource-control friendly.

Reply:
There are three important differences between file::close() and the new set_maximum_connections:
1. You want to wait for a file to be flushed before shutting down, so you need the ability to wait for the close.
2. A close() can fail and you want to report if it failed - need to wait for this too.
3. You can have a thousand files being closed in parallel, and maybe want to not start all the closing operations in parallel.
I think none of these reasons apply in your case - there is no reason (that I can think of) to care about when the closing of **some** of the connections happened, there is no error to report, and even if you set_maximum_connections on a thousand clients (why would you have many of those, anyway?) it won't do anything in parallel - these connections are closed in their natural pace, whenever they finish they just don't get saved.

Anyway, as I said, if you like this future return, I'm not strongly against it - and you already wrote the code. So it's your call. Let me know if you want me to commit this version as-is.

Reply:
Yes, please, commit it as is. If it happens that it doesn't work for us, the client is in experimental namespace and it can be changed easily

Reply:
If this is ture then to be honest, I don't understand why we have file close() at all and why we insist that users wait for it (CC @bhalevy, he's an expert in this).

Anyway, I'll commit this PR. I am guessing that nobody will ever want to wait for this future, but it may very well end up like other famous predictions - like 640 KB of memory will be more than enough for everyone ;-)